File: app_manager/appmanager/src/main/java/app/editors/manager/ui/adapters/diffutilscallback/RecentDiffUtilsCallback.java
package app.editors.manager.ui.adapters.diffutilscallback;

import java.util.List;

import app.documents.core.account.Recent;


public class RecentDiffUtilsCallback extends BaseDiffUtilsCallback<Recent> {

    public RecentDiffUtilsCallback(List<Recent> mNewList, List<Recent> mOldList) {
        super(mNewList, mOldList);
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        if (mOldList.get(oldItemPosition) instanceof Recent && mNewList.get(newItemPosition) instanceof Recent) {
            Recent oldRecent = (Recent) mOldList.get(oldItemPosition);
            Recent newRecent = (Recent) mNewList.get(newItemPosition);
            return oldRecent.getId() == newRecent.getId();
        }
        return false;
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        if (mOldList.get(oldItemPosition) instanceof Recent && mNewList.get(newItemPosition) instanceof Recent) {
            Recent oldRecent = (Recent) mOldList.get(oldItemPosition);
            Recent newRecent = (Recent) mNewList.get(newItemPosition);
            return oldRecent.getDate() == newRecent.getDate();
        }
        return false;
    }
}

File: app_manager/appmanager/src/main/java/app/editors/manager/ui/adapters/diffutilscallback/RecentDiffUtilsCallback.kt
package app.editors.manager.ui.adapters.diffutilscallback

import app.editors.manager.mvp.models.list.Header
import app.editors.manager.mvp.models.ui.RecentUI
import lib.toolkit.base.ui.adapters.holder.ViewType

class RecentDiffUtilsCallback(mNewList: List<ViewType>, mOldList: List<ViewType>) :
    BaseDiffUtilsCallback<ViewType>(mNewList, mOldList) {

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return if (mOldList[oldItemPosition] is Header && mNewList[newItemPosition] is Header) {
            (mOldList[oldItemPosition] as Header).title == (mNewList[newItemPosition] as Header).title
        } else if (mOldList[oldItemPosition] is RecentUI && mNewList[newItemPosition] is RecentUI) {
            (mOldList[oldItemPosition] as RecentUI).id == (mNewList[newItemPosition] as RecentUI).id
        } else false
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return if (mOldList[oldItemPosition] is RecentUI && mNewList[newItemPosition] is RecentUI)
            return (mOldList[oldItemPosition] as RecentUI).date == (mNewList[newItemPosition] as RecentUI).date
        else false
    }
}